src/orion/core/utils/tests.py

```diff
@@ -0,0 +1,212 @@
+.. module:: state
+   :platform: Unix
+   :synopsis: Encapsulate orion state
+
+"""
+
+import datetime
+import os
+import tempfile
+
+import yaml
+
+from orion.core.io.database import Database
+from orion.core.io.database.ephemeraldb import EphemeralDB
+from orion.core.io.database.mongodb import MongoDB
+from orion.core.io.database.pickleddb import PickledDB
+from orion.core.utils import SingletonAlreadyInstantiatedError
+from orion.core.worker.experiment import Experiment
+from orion.core.worker.trial import Trial
+from orion.storage.base import get_storage, Storage
+from orion.storage.legacy import Legacy
+
+
+def _remove(file_name):
+    try:
+        os.remove(file_name)
+    except FileNotFoundError:
+        pass
+
+
+def _select(lhs, rhs):
+    if lhs:
+        return lhs
+    return rhs
+
+
+def default_datetime():
+    """Return default datetime"""
+    return datetime.datetime(1903, 4, 25, 0, 0, 0)
+
+
+class MockDatetime(datetime.datetime):
+    """Fake Datetime"""
+
+    @classmethod
+    def utcnow(cls):
+        """Return our random/fixed datetime"""
+        return default_datetime()
+
+
+def _get_default_test_database():
+    """Return default configuration for the test database"""
+    _, filename = tempfile.mkstemp('orion_test')
+
+    return {
+        'storage_type': 'legacy',
+        'type': 'PickledDB',
+        'host': filename
+    }
+
+
+# pylint: disable=no-self-use,protected-access
+class OrionState:
+    """Setup global variables and singleton for tests
+
+    it swaps the singleton with none at startup and restore them after the tests.
+    It also initialize PickleDB as the storage for testing.
+    We use PickledDB as our storage mock
+
+    Parameters
+    ----------
+    config: YAML
+        YAML config to apply for this test
+
+    experiments: list
+        List of experiments to insert into the database
+
+    trials: list
+        List of trials to insert into the database
+
+    workers: list
+        List of workers to insert into the database
+
+    resources: list
+        List of resources to insert into the database
+
+    database: dict
+        Configuration of the underlying database
+
+    Examples
+    --------
+    >>> myconfig = {...}
+    >>> with OrionTestState(myconfig):
+        ...
+
+    """
+
+    # TODO: Fix these singletons to remove Legacy, MongoDB, PickledDB and EphemeralDB.
+    SINGLETONS = (Storage, Legacy, Database, MongoDB, PickledDB, EphemeralDB)
+    singletons = {}
+    database = None
+    experiments = []
+    trials = []
+    resources = []
+    workers = []
+    tempfile = None
+
+    def __init__(self, experiments=None, trials=None, workers=None, resources=None,
+                 from_yaml=None, database=None):
+        if from_yaml is not None:
+            with open(from_yaml) as f:
+                exp_config = list(yaml.safe_load_all(f))
+                experiments = exp_config[0]
+                trials = exp_config[1]
```

Comment:
I suggest we get the list of experiments and trials through a key inside the yaml, for example:
```
experiments:
                      [
                      ...
                      ]
trials
                      [
                      ...
                      ]
```
This way we could have a yaml file of only trials without having experiments.

Reply:
I agree, the reason why it is setup like this is because that is how the current YAMLs are structed. We might have to support both in the meantime. maybe something like this ?

```
    def from_yaml(self, file_name):
        """Load a YAML files containing the initial state of the database"""
        data = {
            'experiments': [],
            'trials': []
        }

        if file_name is None:
            return data

        with open(file_name) as f:
            try:
                # try to load the YAML as a dictionary
                return data.update(dict(yaml.safe_load_all(f)))

            except ValueError:
                # TODO: Remove legacy loading and use yaml dictionary instead
                # if it fails try the legacy loading scheme
                exp_config = list(yaml.safe_load_all(f))
                data['experiments'] = exp_config[0]
                data['trials'] = exp_config[1]
                return data
```